RandLAPACK/drivers/rl_cqrrp_gpu.hh

```diff
@@ -0,0 +1,804 @@
+            h_work_getrf = h_work_getrf_vector.data();
+        }
+        lapack::getrf(cols, sampling_dimension, A_sk_trans, n, J_buffer_lu, d_work_getrf, d_size_getrf, h_work_getrf, h_size_getrf, d_info, lapack_queue);
+        // Fill the pivot vector, apply swaps found via lu on A_sk'.
+        RandLAPACK::cuda_kernels::LUQRCP_piv_porcess_gpu(strm, sampling_dimension, cols, J_buffer, J_buffer_lu);
+        
+        if(this -> timing) {
+            cudaStreamSynchronize(strm);
+            nvtxRangePushA("copy_A_sk");
+            copy_A_sk_t_start = high_resolution_clock::now();
+        }
+        // Instead of copying A_sk_work into A_sk_copy_col_swap, we ``swap'' the pointers.
+        // This is safe, as A_sk is not needed outside of ICQRRP.
+        std::swap(A_sk_copy_col_swap_work, A_sk_work);
+        
+        if(this -> timing) {
+            cudaStreamSynchronize(strm);
+            nvtxRangePop();
+            copy_A_sk_t_stop = high_resolution_clock::now();
+            copy_A_sk_t_dur += duration_cast<microseconds>(copy_A_sk_t_stop - copy_A_sk_t_start).count();
+            nvtxRangePushA("piv_A_sk");
+            qrcp_piv_t_start = high_resolution_clock::now();
+        }
+
+        // Apply pivots to A_sk
+        RandLAPACK::cuda_kernels::col_swap_gpu(strm, sampling_dimension, cols, cols, A_sk_work, d, A_sk_copy_col_swap_work, d, J_buffer);
+
+        if(this -> timing) {
+            cudaStreamSynchronize(strm);
+            nvtxRangePop();
+            qrcp_piv_t_stop = high_resolution_clock::now();
+            qrcp_piv_t_dur += duration_cast<microseconds>(qrcp_piv_t_stop - qrcp_piv_t_start).count();
+        }
+
+        // Perform an unpivoted QR on A_sk
+        if(iter == 0) {
+            lapack::geqrf_work_size_bytes(sampling_dimension, cols, A_sk_work, d, &d_size_geqrf, &h_size_geqrf, lapack_queue);
+            d_work_geqrf = blas::device_malloc< char >( d_size_geqrf, lapack_queue );
+            std::vector<char> h_work_geqrf_vector( h_size_geqrf );
+            h_work_geqrf = h_work_geqrf_vector.data();
+        }
+        lapack::geqrf(sampling_dimension, cols, A_sk_work, d, Work2, d_work_geqrf, d_size_geqrf, h_work_geqrf, h_size_geqrf, d_info, lapack_queue);
+        
+        if(this -> timing) {
+            lapack_queue.sync();
+            nvtxRangePop();
+            qrcp_t_stop = high_resolution_clock::now();
+            qrcp_t_dur += duration_cast<microseconds>(qrcp_t_stop - qrcp_t_start).count();
+            nvtxRangePushA("copy_A");
+            copy_A_t_start = high_resolution_clock::now();
+        }
+        // THIS IS ALREADY DONE ABOVE
+        // Need to premute trailing columns of the full R-factor.
+        // Remember that the R-factor is stored the upper-triangular portion of A.
+        // Pivoting the trailing R and the ``current'' A.      
+        //RandLAPACK::cuda_kernels::copy_mat_gpu(strm, m, cols, &A[lda * curr_sz], lda, A_copy_col_swap, lda, false);    
```

Comment:
So ... can you remove this commented-out code?

Reply:
Yeah, i just wanted to signify that if the copy was not done on a separate stream, we should have done it here.